packages/gatsby/src/commands/serve.js

```diff
@@ -1,13 +1,39 @@
 /* @flow weak */
 const path = require(`path`)
 const openurl = require(`better-opn`)
+const fs = require(`fs-extra`)
 const signalExit = require(`signal-exit`)
 const compression = require(`compression`)
 const express = require(`express`)
-const historyFallback = require(`express-history-api-fallback`)
 const getConfigFile = require(`../bootstrap/get-config-file`)
 const preferDefault = require(`../bootstrap/prefer-default`)
 const chalk = require(`chalk`)
+const { match: reachMatch } = require(`@reach/router/lib/utils`)
+
+const getPages = directory =>
+  fs
+    .readFile(path.join(directory, `.cache`, `pages.json`))
+    .then(contents => JSON.parse(contents))
+    .catch(() => [])
+
+const historyRouter = (pages, options) => {
+  const clientOnlyRoutes = pages
+    .filter(page => page.matchPath)
+    .map(page => page.matchPath)
+  return (req, res, next) => {
+    const { url } = req
+    if (clientOnlyRoutes.some(route => reachMatch(route, url) !== null)) {
+      if (req.accepts(`html`)) {
```

Comment:
maybe we could move this check up so we don't even iterate over routes if it's not "html" request?

Reply:
Yup!